### PR TITLE
cloudwatchevent_rule: Fix json input handling for input_template

### LINKED ISSUE #1842: Regression: cloudwatchevent_rule doesn't properly support JSON input_templates
### Summary

I reported this same bug over in the [community.aws collection](https://github.com/ansible-collections/community.aws/issues/1348) last year.  It seems to have returned [in this code](https://github.com/ansible-collections/amazon.aws/blob/main/plugins/modules/cloudwatchevent_rule.py#L444).

Please see the [previous issue report](https://github.com/ansible-collections/community.aws/issues/1348) for a detailed reproduction example.

### Issue Type

Bug Report

### Component Name

cloudwatchevent_rule

### Ansible Version

```console (paste below)
$ ansible --version

```

### Collection Versions

```console (paste below)
$ ansible-galaxy collection list
```

### AWS SDK versions

```console (paste below)
$ pip show boto boto3 botocore
```

### Configuration

```console (paste below)
$ ansible-config dump --only-changed

```

### OS / Environment

_No response_

### Steps to Reproduce

Here is a working example using the AWS CLI:

```
eventbridge_target.json
{
   "Rule": "myRule",
   "Targets": [{
      ...
      "InputTransformer": {
         "InputPathsMap": {
            "s3Key": "$.detail.object.key"
         },
         "InputTemplate": "{\"containerOverrides\": [{\"name\": \"thrivSportsJob-stage_jobmg10\", \"environment\": [{\"name\": \"S3_KEY\",\"value\": <s3Key>}]}]}"
      }
   }]
}
```

Which can be executed with a `aws events put-targets --cli-input-json file://eventbridge_target.json`

However, when parsing the target configuration, the input is wrapped in a string:
```
                target_request['InputTransformer']['InputTemplate'] = '"{0}"'.format(
                    target['input_template']
                )
```

AWS will not route an event with an invalid transformer template. So the input for a JSON template must begin with a `{`, not a `"`. My [solution](https://github.com/alalonde/community.aws/commit/33e57f964687c09e81de238f936fc5b63f95bd76) was to remove the string wrapping, AND prepend the template with a space, as [recommended here](https://stackoverflow.com/a/32014283/331791). This results in:

```yaml
      -  name: Define trigger input template
         set_fact:
            trigger_input_template: |
               {
                  "containerOverrides": [{
                     "name": "{{container_name}}",
                     "environment": [{
                        "name": "S3_KEY",
                        "value": <s3Key>
                     }]
                  }]
               }
      -  name: Create rule in EventBridge
         community.aws.cloudwatchevent_rule:
            ...
            targets:
               -  id: "myJob-{{env}}"
                  ...
                  input_paths_map:
                     s3Key: "$.detail.object.key"
                  input_template: " {{trigger_input_template}}"
                  ...
```

### Expected Results

I expected the input template to be wrapped by `{ ... }` but was instead `"{ ... }"`

image

### Actual Results

```console (paste below)

```

### Code of Conduct

- [X] I agree to follow the Ansible Code of Conduct

### PR DESCRIPTION
**Pull request by mandar242**
##### SUMMARY

Fixes #1842 

This PR
- Moves `_snakify()` out of `class CloudWatchEventRule` https://github.com/ansible-collections/amazon.aws/pull/1883/files#diff-3a2f223edc4e34ea28b9859827a830844d92a8cd97f1300a2d16b1703a083bc8R213-R215

- Fixes conditional in function `_targets_to_put()` to avoid wrong results due to `_snakify()` https://github.com/ansible-collections/amazon.aws/pull/1883/files#diff-3a2f223edc4e34ea28b9859827a830844d92a8cd97f1300a2d16b1703a083bc8R461-R463

- Updates code to add quotes `"` to `input_template` only when the given input is not JSON to avoid quotes being added to provided JSON input https://github.com/ansible-collections/amazon.aws/pull/1883/files#diff-3a2f223edc4e34ea28b9859827a830844d92a8cd97f1300a2d16b1703a083bc8R452-R458

##### ISSUE TYPE

- Bugfix Pull Request

##### COMPONENT NAME

cloudwatchevent_rule
##### ADDITIONAL INFORMATION

```paste below

```

### CODE REVIEW

**Review by abikouo — COMMENTED** [review]
Would be nice to add some unit tests for this update

**Comment by mandar242** [comment]
> Would be nice to add some unit tests for this update

there are integration tests for `cloudwatchevent_rule`, should we be adding more tests to existing integration tests or create new unit tests for this change?

**Review by abikouo — APPROVED** [review]

File: plugins/modules/cloudwatchevent_rule.py [hunk 1]
```diff
@@ -204,6 +204,19 @@ def _format_json(json_string):
...
 
+def _validate_json(s):
+    try:
+        json.loads(s)
+        return True
+    except json.JSONDecodeError:
+        return False
+
+
+def _snakify(dict):
```
Comment (hakbailey) [inline plugins/modules/cloudwatchevent_rule.py:215]: Instead of moving this here, could we just call `camel_dict_to_snake_dict` directly wherever this is called?

**Review by hakbailey — COMMENTED** [review]
I think it would be good to add both an integration test and an example in the module docs for a json input template.

**Comment by tremble** [comment]
> > Would be nice to add some unit tests for this update
> 
> there are integration tests for `cloudwatchevent_rule`, should we be adding more tests to existing integration tests or create new unit tests for this change?

Personally I'd stick to the integration tests for this change.  The code is only consumed by this module, so adding unit tests is IMO "busy-work" that brings minimal value over the top of the integration tests.  The only exception would be if they're handling hard to trigger edge cases.

File: plugins/modules/cloudwatchevent_rule.py (hunk 1 above)
Comment (alinabuzachis) [reply plugins/modules/cloudwatchevent_rule.py:215]: Agree!

Comment (mandar242) [reply plugins/modules/cloudwatchevent_rule.py]: removed _snakify()

**Review by hakbailey — APPROVED** [review]

**Comment by alalonde** [comment]
When will this be released?
